pkg/codegen/internal/test/testdata/simple-methods-schema/python/pulumi_example/foo.py

```diff
@@ -0,0 +1,142 @@
+                 opts: Optional[pulumi.ResourceOptions] = None,
+                 __props__=None):
+        """
+        Create a Foo resource with the given unique name, props, and options.
+        :param str resource_name: The name of the resource.
+        :param pulumi.ResourceOptions opts: Options for the resource.
+        """
+        ...
+    @overload
+    def __init__(__self__,
+                 resource_name: str,
+                 args: Optional[FooArgs] = None,
+                 opts: Optional[pulumi.ResourceOptions] = None):
+        """
+        Create a Foo resource with the given unique name, props, and options.
+        :param str resource_name: The name of the resource.
+        :param FooArgs args: The arguments to use to populate this resource's properties.
+        :param pulumi.ResourceOptions opts: Options for the resource.
+        """
+        ...
+    def __init__(__self__, resource_name: str, *args, **kwargs):
+        resource_args, opts = _utilities.get_resource_args_opts(FooArgs, pulumi.ResourceOptions, *args, **kwargs)
+        if resource_args is not None:
+            __self__._internal_init(resource_name, opts, **resource_args.__dict__)
+        else:
+            __self__._internal_init(resource_name, *args, **kwargs)
+
+    def _internal_init(__self__,
+                 resource_name: str,
+                 opts: Optional[pulumi.ResourceOptions] = None,
+                 __props__=None):
+        if opts is None:
+            opts = pulumi.ResourceOptions()
+        if not isinstance(opts, pulumi.ResourceOptions):
+            raise TypeError('Expected resource options to be a ResourceOptions instance')
+        if opts.version is None:
+            opts.version = _utilities.get_version()
+        if opts.id is None:
+            if __props__ is not None:
+                raise TypeError('__props__ is only valid when passed in combination with a valid opts.id to get an existing resource')
+            __props__ = FooArgs.__new__(FooArgs)
+
+        super(Foo, __self__).__init__(
+            'example::Foo',
+            resource_name,
+            __props__,
+            opts)
+
+    @staticmethod
+    def get(resource_name: str,
+            id: pulumi.Input[str],
+            opts: Optional[pulumi.ResourceOptions] = None) -> 'Foo':
+        """
+        Get an existing Foo resource's state with the given name, id, and optional extra
+        properties used to qualify the lookup.
+
+        :param str resource_name: The unique name of the resulting resource.
+        :param pulumi.Input[str] id: The unique provider ID of the resource to lookup.
+        :param pulumi.ResourceOptions opts: Options for the resource.
+        """
+        opts = pulumi.ResourceOptions.merge(opts, pulumi.ResourceOptions(id=id))
+
+        __props__ = FooArgs.__new__(FooArgs)
+
+        return Foo(resource_name, opts=opts, __props__=__props__)
+
+    @pulumi.output_type
+    class BarResult:
```

Comment:
There are a few different options on where to emit the result class for methods. I considered emitting them in the same module as the resource, prefixed with the resource name (e.g. `FooBarResult` for a resource `Foo`). I also considered putting them in some other module like `results` (similar to what we do for output types with `outputs`). In the end, nesting the result classes within the resource class felt best (e.g. `Foo.BarResult`), which is what I've done here. This avoids adding more noise to the module for types that you'll rarely (if ever) need to refer to directly.

Reply:
Sounds good to me. In what circumstances might you need to refer to these types directly?

Reply:
> In what circumstances might you need to refer to these types directly?

I can only really come up with contrived examples where maybe you declare a variable with a type hint ahead of calling the method and assigning the result to the variable at a later point.

```python
result: Optional[pulumi.Output[Foo.BarResult]] = None
# ... sometime later ...
result = foo.bar() 
```

The vast majority of the time you would be calling the method and using the result directly, with no need to refer to the type because it'd be inferred from the return type hint on the method:

```python
result = foo.bar()
```